test/EngineTest.cpp

```diff
@@ -23,114 +32,259 @@ auto I = [](const auto& id) {
+    const IdTable& sampleSolution,
+    const bool testForSorted = false,
+    const size_t jc = 0,
+    ad_utility::source_location t = ad_utility::source_location::current()
+    ) {
+  // For generating more informative messages, when failing the comparison.
+  std::stringstream traceMessage;
+  traceMessage << "compareIdTableWithSolution comparing IdTable\n";
+  for (size_t row = 0; row < table.size(); row++) {
+    for (size_t column = 0; column < table.numColumns(); column++) {
+      traceMessage << table(row, column) << " ";
+    }
+    traceMessage << "\n";
+  }
+  traceMessage << "with IdTable \n";
+  for (size_t row = 0; row < sampleSolution.size(); row++) {
+    for (size_t column = 0; column < sampleSolution.numColumns(); column++) {
+      traceMessage << sampleSolution(row, column) << " ";
+    }
+    traceMessage << "\n";
+  }
+  auto trace = generateLocationTrace(t, traceMessage.str());
+ 
+  // Because we compare tables later by sorting them, so that every table has
+  // one definit form, we need to create local copies.
+  IdTable localTable = table.clone();
+  IdTable localSampleSolution = sampleSolution.clone();
+
+  // Do the IdTable and sampleSolution have the same dimensions?
+  ASSERT_EQ(localTable.size(), localSampleSolution.size());
+  ASSERT_EQ(localTable.numColumns(), localSampleSolution.numColumns());
+
+  if (testForSorted) {
+    // Is the table sorted by join column?
+    auto oldEntry = localTable(0, jc);
+    for (size_t i = 1; i < localTable.size(); i++) {
+      ASSERT_TRUE(oldEntry <= localTable(i, jc));
+      oldEntry = localTable(i, jc);
+    }
+  }
+
+  // Sort both the table and the sampleSolution, so that both have a definite
+  // form for comparison.
+  auto sortFunction = [](
+      const auto& element1,
+      const auto& element2)
+    {
+      size_t i = 0;
+      while (i < (element1.numColumns() - 1) && element1[i] == element2[i]) {
+        i++;
+      }
+      return element1[i] < element2[i];
+    };
+  std::sort(localTable.begin(), localTable.end(), sortFunction);
+  std::sort(localSampleSolution.begin(), localSampleSolution.end(), sortFunction);
+
+  ASSERT_EQ(localTable, localSampleSolution);
+}
+
+/*
+ * @brief Joins two IdTables togehter with the given join function and returns
+ * the result.
+ *
+ * @tparam JOIN_FUNCTION is used to allow the transfer of any type of
+ *  lambda function, that could contain a join function. You never have to
+ *  actually specify this parameter , just let inference do its job.
+ *
+ * @param tableA, tableB the tables. 
+ * @param jcA, jcB are the join columns for the tables.
+ * @param func the function, that will be used for joining the two tables
+ *  together. Look into src/engine/Join.h for how it should look like.
+ *
+ * @returns tableA and tableB joined together in a IdTable.
+ */
+template<typename JOIN_FUNCTION>
+IdTable useJoinFunctionOnIdTables(
+        const IdTable& tableA,
+        const size_t jcA,
```

Comment:
Something even nicer we need to start doing:
make a struct `IdTableAndJoinColumn` and use it everywhere where we have currently something like `tableA, jcA, tableB, jcB`.